JDK-8293364: IGV: Refactor Action in EditorTopComponent and fix minor bugs

tobiasholenstein opened the pull request:
Refactor the Actions in EditorTopComponent (com.sun.hotspot.igv.view.actions). Move Action specific code from EditorTopComponent to the corresponding Action.

# Refactoring of com.sun.hotspot.igv.view.actions and EditorTopComponent
- Created a new `ExportGraph` Action and moved corresponding functions `exportToSVG(..)` and `exportToPDF(..)` to new `ExportGraph.java`
- Moved key bindings for satellite-view (pressing S) from `EditorTopComponent` to `OverviewAction.java` 
- Moved Action specific code from `EditorTopComponent` to the corresponding `XXXAction.java`
- Changed `PrevDiagramAction`, `ExpandDiffAction`, `ExtractAction`, `HideAction`, `NextDiagramAction`, `ReduceDiffAction` and `ShowAllAction` to be context aware `ContextAction<DiagramViewModel>` actions and use more modern `@ActionRegistration` to move away from manually defining actions in `layer.xml`   
- new `addContextListener` / `removeContextListener` function  in `ContextAction<T>` enables context aware actions to define to which `ChangedEvent` they want to react to

# Fixing minor Bugs
- "Show empty blocks in control-flow graph view" is selected by default but only enabled in CFG view. 
This is distracting for the eye when we are not in CFG:
<img width="192" alt="cfg_before" src="https://user-images.githubusercontent.com/71546117/188463472-b928075f-a687-46d6-a871-35f287e943c7.png">
Now "Show empty blocks in control-flow graph view" is not selected anymore when disabled (greyed out)
<img width="192" alt="cfg_node_disable" src="https://user-images.githubusercontent.com/71546117/188463809-c998b03e-cc2f-48c1-8f5f-135edffa9c8c.png">
But still gets selected by default when enabled
<img width="197" alt="cfg_now" src="https://user-images.githubusercontent.com/71546117/188463853-33e6e7cb-b8ff-4c75-a522-63fd858b42ef.png">

- "Extract current set of selected nodes", "Hide selected nodes" and "show all nodes" were always enabled, even when they didn't effect anything. 
<img width="99" alt="selection_before" src="https://user-images.githubusercontent.com/71546117/188464555-45ef06af-b247-46b0-aaa2-4c2397566f8c.png">
Now "Extract current set of selected nodes", "Hide selected nodes" are disabled (greyed out) when no nodes are selected. And "show all nodes" is disabled (greyed out) when all nodes are already visible.
<img width="91" alt="selection_now" src="https://user-images.githubusercontent.com/71546117/188464589-61310b13-ba58-45a9-b2d8-d3be0da582ab.png">

- "Reduce the difference selection" got stuck when at the last graphs in the group because it got greyed out. 
<img width="130" alt="reduce_stuck" src="https://user-images.githubusercontent.com/71546117/188465260-6fca7ca6-3e8e-4bd9-9b70-4ac19d79eff5.png">
duce the difference selection" 
Now "Reduce the difference selection" works as expected:
<img width="152" alt="reduce_now" src="https://user-images.githubusercontent.com/71546117/188465314-bce441ba-20d6-4e3d-ba4e-ff1f4a79cd9d.png">

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293364](https://bugs.openjdk.org/browse/JDK-8293364): IGV: Refactor Action in EditorTopComponent and fix minor bugs


### Reviewers
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**) ⚠️ Review applies to [b78a8fb5](https://git.openjdk.org/jdk/pull/10170/files/b78a8fb5da77ac2a8bd5b21803c88f24e0b6ddcb)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [b78a8fb5](https://git.openjdk.org/jdk/pull/10170/files/b78a8fb5da77ac2a8bd5b21803c88f24e0b6ddcb)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10170/head:pull/10170` \
`$ git checkout pull/10170`

Update a local copy of the PR: \
`$ git checkout pull/10170` \
`$ git pull https://git.openjdk.org/jdk pull/10170/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10170`

View PR using the GUI difftool: \
`$ git pr show -t 10170`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10170.diff">https://git.openjdk.org/jdk/pull/10170.diff</a>

</details>

robcasloz reviewed (changes requested):
This changeset seems to disable the keyboard shortcuts for `Extract`, `Show all nodes`, and `Hide` right after a graph is opened. Interestingly, after clicking around for a while, the keyboard shortcuts start working again. Please let me know if you need more details to reproduce the problem, hopefully it is reproducible in other platforms than my own (Ubuntu 20.04).

tobiasholenstein commented:
> This changeset seems to disable the keyboard shortcuts for `Extract`, `Show all nodes`, and `Hide` right after a graph is opened. Interestingly, after clicking around for a while, the keyboard shortcuts start working again. Please let me know if you need more details to reproduce the problem, hopefully it is reproducible in other platforms than my own (Ubuntu 20.04).

Thanks for spotting that @robcasloz! Defining the `Shortcuts` in `layer.xml` is not recommended anymore and cause problems with `ContextAction`. The more modern approach is to use `@ActionXXX` annotations and define the shortcuts there.

Also, `ContextAction<DiagramViewModel>` listened only to `getDiagramChangedEvent()`. I changed them to also listened to `getViewChangedEvent()`, `getViewPropertiesChangedEvent()` and `getHiddenNodesChangedEvent()`. Otherwise the context aware action missed being updated when node selection or view changed. 

 I updated to PR description accordingly

robcasloz reviewed (approved):
I have tested the changeset again and did not find any issues, keyboard shortcuts and contextual enabling/disabling of toolbar buttons works as expected. I only have a few style comments, please consider addressing them before integration.

tobiasholenstein commented on src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/actions/ImportAction.java:
```diff
@@ -31,26 +31,25 @@
-import org.openide.util.actions.CallableSystemAction;
+import org.openide.util.RequestProcessor;
 import org.openide.util.Utilities;
+import org.openide.util.actions.CallableSystemAction;
```
I agree, if the file is not touched, imports should not be modified. Hopefully after https://github.com/openjdk/jdk/pull/10197 this mistake shouldn't happen anymore :)

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/EditorTopComponent.java:
```diff
@@ -676,50 +413,20 @@ public UndoRedo getUndoRedo() {
-    }
+        setLayout(new java.awt.BorderLayout());
+
+    }// </editor-fold>//GEN-END:initComponents
```
I reverted the change. The NetBeans' Form Editor opens it, but just shows an empty plane. I don't really know why there is an `JCheckBox jCheckBox1` if we don't use it. But let's just leave it the way NetBeans generated it.

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/ExportAction.java:
```diff
@@ -25,31 +25,43 @@
-/**
- *
- * @author Thomas Wuerthinger
- */
```
Done

chhagedorn reviewed (changes requested):
I only have some code style specific comments. Otherwise, good cleanup!

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/DiagramViewModel.java:
```diff
@@ -183,15 +183,9 @@ public void setShowEmptyBlocks(boolean b) {
-        hideDuplicates = b;
         InputGraph currentGraph = getFirstGraph();
-        if (hideDuplicates) {
+        if (b) {
```
done

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/EditorTopComponent.java:
```diff
@@ -409,62 +298,16 @@ public void zoomIn() {
-    @Override
-    protected String preferredID() {
-        return PREFERRED_ID;
+        TopComponent topComponent = EditorTopComponent.getRegistry().getActivated();
```
done

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/EnableBlockLayoutAction.java:
```diff
@@ -24,21 +24,29 @@
-public class EnableBlockLayoutAction extends AbstractAction {
+public class EnableBlockLayoutAction extends AbstractAction implements PropertyChangeListener {
+
+    EditorTopComponent editor;
```
done

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/ReduceDiffAction.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Override
+    public Action createContextAwareInstance(Lookup actionContext) {
+        return this;
+    }
```
Unfortunately not, we need to override it because we extend `CallableSystemAction`

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/EditorTopComponent.java:
```diff
@@ -321,84 +255,39 @@ public void changed(DiagramViewModel source) {
-    public DiagramViewModel getDiagramModel() {
-        return rangeSliderModel;
+    public DiagramViewModel getModel() {
+        return  scene.getModel();
```
done

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/EditorTopComponent.java:
```diff
@@ -321,84 +255,39 @@ public void changed(DiagramViewModel source) {
+        return getModel().getDiagramToView();
+    }
 
+    public void selectionMode(boolean b) {
```
done

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/ReduceDiffAction.java:
```diff
@@ -0,0 +1,105 @@
+        model.getDiagramChangedEvent().removeListener(this);
+        model.getViewPropertiesChangedEvent().removeListener(this);
+        model.getHiddenNodesChangedEvent().removeListener(this);
+    }
```
Done. I added an abstract superclass `ModelAwareAction.java`

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/EditorTopComponent.java:
```diff
@@ -505,151 +331,62 @@ public void setSelectedFigures(List<Figure> list) {
-        getModel().showNot(new HashSet<Integer>());
-    }
+    @Override
+    public void componentOpened() { }
```
I removed it

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/ExpandPredecessorsAction.java:
```diff
@@ -37,7 +41,28 @@ public final class ExpandPredecessorsAction extends CallableSystemAction {
         EditorTopComponent editor = EditorTopComponent.getActive();
         if (editor != null) {
-            editor.expandPredecessors();
+            Set<Figure> oldSelection = editor.getModel().getSelectedFigures();
```
I agree. I created a new `ExpandAdjacentAction` superclass now

tobiasholenstein commented:
@chhagedorn I integrated your suggestions now

chhagedorn reviewed (approved):
Thanks for doing the suggested updates. Apart from two small things, it looks good to me!

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/EnableBlockLayoutAction.java:
```diff
@@ -23,42 +23,30 @@
+public class EnableBlockLayoutAction extends EnableLayoutAction {
 
-    EditorTopComponent editor;
+    private final EditorTopComponent editor;
```
done

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/ExpandAdjacentAction.java:
```diff
@@ -0,0 +1,81 @@
+        List<Figure> get(Figure f);
+    }
+
+    protected void expandFigures(AdjacentOperator adjacentOp) {
```
done

tobiasholenstein commented:
@chhagedorn and @robcasloz thanks for the reviews!

/integrate
